### PR TITLE
LoRA Trainer: `train_only_after` option to control which part of your input to train on

### PR DESCRIPTION
**Pull request by mcmonkey4eva**
tldr: adds new option `train_only_after` under Advanced that you can fill in a string and it will ignore text before that part and only count text after it for training.

Relevant backend `labels` with its magic value `-1` is documented @ https://huggingface.co/transformers/v3.1.0/custom_datasets.html

Labels seems to have other abilities that might be worth looking into? Not sure.

This feature was shown to be in use in the original Stanford Alpaca https://github.com/tatsu-lab/stanford_alpaca/blob/main/train.py#L123 but to my knowledge hasn't been ported to other trainers (eg the original alpaca-lora that was the starting point for this code, way back when, in the ancient times). Not sure if that means it's not that useful, or just nobody's really noticed it's missing before now.

So for training on an Alpaca formatted dataset, you can set this to `### Response:` 

Operates in string space because token space is cursed for some theoretically valid matchers (ie anything that lacks spaces).

The code is a bit longer than I'd like it to be, because operating in string space means replicating the padding code (as otherwise it would double-pad), but not too bad.

I tested and verified that training with/without the setting works as normal, and that the splitter does exactly what it's meant to, and that the labels are being received and processed by the HF Trainer.
I have not tested the effect on quality of training results with vs without.

---

Also, additional fixes:

- Fixed last tokens being stripped. Not sure if this was happening before or as a byproduct of the encoder call change, but either way, it definitely doesn't wrongly strip now!
- HF datasets are cursedbad. See prior issue <https://github.com/oobabooga/text-generation-webui/discussions/1586> but also new issue - they give errors when loading from cache when new labels data is present it seems. ... also after disabling that, it loads so much faster. Wow. Ow.

### CODE REVIEW

File: modules/training.py [hunk 1]
```diff
@@ -245,11 +247,35 @@ def do_train(lora_name: str, always_override: bool, save_steps: int, micro_batch
...
+            labels = [1] * len(input_ids)
+
+        else:
+            ind = prompt.index(train_only_after) + len(train_only_after)
+            before_tokens = encode(prompt[:ind])[:-1]
+            after_tokens = encode(prompt[ind:])[1:]
+
+            full_length = len(after_tokens) + len(before_tokens)
+            if full_length > cutoff_len:
+                after_tokens = after_tokens[:cutoff_len - len(before_tokens)]
```
Comment (oobabooga) [inline modules/training.py:267]: Could there be an edge case where `len(before_tokens) > cutoff_len`?

**Review by oobabooga — COMMENTED** [review]
This looks excellent!

File: modules/training.py [hunk 2]
```diff
@@ -245,11 +247,35 @@ def do_train(lora_name: str, always_override: bool, save_steps: int, micro_batch
...
+
+        if train_only_after == '' or train_only_after not in prompt:
+            input_ids = encode(prompt)
+            input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
+            labels = [1] * len(input_ids)
+
+        else:
+            ind = prompt.index(train_only_after) + len(train_only_after)
+            before_tokens = encode(prompt[:ind])[:-1]
+            after_tokens = encode(prompt[ind:])[1:]
```
[the remark this answers is not in the document]
Comment (mcmonkey4eva) [reply modules/training.py:263]: Oops - good catch on checking that, I had fixed it for the non-onlyafter case but not here. There is no eos token generated, it's a bos that needs to be stripped. I've only tested LLaMA - I *assume* every tokenizer emits one, I haven't verified that though.
Can toss `print(f"[Debug] before ({shared.tokenizer.decode(before_tokens)}) after ({shared.tokenizer.decode(after_tokens)})")` after the `after_tokens = ` line to see exactly what it's generating if in doubt.

File: modules/training.py (hunk 1 above)
Comment (mcmonkey4eva) [reply modules/training.py:267]: No, that's guaranteed against on line 251 `max_length=cutoff_len` in the tokenizer encode call.

File: modules/training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcmonkey4eva) [reply modules/training.py]: Formal guarantee doesn't hurt, added. Fits cleanly.

**Comment by oobabooga** [comment]
Thank you for the fixes @mcmonkey4eva !

File: modules/training.py [hunk 3]
```diff
@@ -245,11 +247,38 @@ def do_train(lora_name: str, always_override: bool, save_steps: int, micro_batch
...
-        result = shared.tokenizer(prompt, truncation=True, max_length=cutoff_len + 1, padding="max_length")
+
+        if train_only_after == '' or train_only_after not in prompt:
+            input_ids = encode(prompt, True)
+            input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
+            labels = [1] * len(input_ids)
+
+        else:
+            ind = prompt.index(train_only_after) + len(train_only_after)
+            before_tokens = encode(prompt[:ind], False)
```
Comment (oobabooga) [inline modules/training.py:265]: Just one more question, why remove the bos token from before _tokens?

Comment (mcmonkey4eva) [reply modules/training.py:265]: Wait you're right good question. I think the answer is I need proper sleep before submitting PRs.
That probably shouldn't be stripped.

Comment (oobabooga) [reply modules/training.py:265]: Heh. I have added it back here: https://github.com/oobabooga/text-generation-webui/commit/63ce5f9c28829a24be2d0eb3f2e57c222c14c042

(if you later conclude that no, we don't want that bos token, let me know)
